---
title: "MRCAUSE_workflow_clumped_oilyfish_GCST90132314_CVD_Aragam"
author: "KJS"
date: "2024-09-17"
output: html_document
---

About diet data: https://kp4cd.org/sites/default/files/READMEs/Cole_UKB_Diet_GWAS_README.pdf
* GRCh37/hg19
* ALLELE1 = first allele in UKB bim file, effect allele
* ALLELE0 = second allele in UKB bim file, non-effect allele

```{r Set up}
# Set wd
wdir <- "/pl/active/colelab/users/kjames/enviroMR/"
setwd(wdir)

# Libraries
library(readr)
library(dplyr)
library(cause)
library(ggplot2)
```

```{r Get data}
# Set traits for naming files 
exposure <- "whiteXwholegrain"
outcome <- "GCST90132314_CVD_Aragam" 
EXP_name <- "White vs Whole Grain or Brown Bread"
OUT_name <- "Cardiovascular Disease"

# Load exposure data
X1 <- read.table(gzfile("/pl/active/colelab/common/published_gwas/diet/cole_naturecomm_2020/BOLTlmm_UKB_genoQCEURN455146_v3_diet_bread_typeused.1448.average_bin4_BgenSnps_mac20_maf0.005_info0.6.gz"), header = TRUE)
head(X1)

# Load outcome data
X2 <- readRDS("/pl/active/colelab/users/kjames/enviroMR/interim_data/outcome/formatted_GCST90132314_CVD_Aragam_GRCh37.rds")
head(X2)
str(X2)
```

```{r Step 1 - Merge}
# Get X1-X2 combined data
# FYI - from the vignette, https://jean997.github.io/cause/ldl_cad.html 
# gwas_merge will merge two data sets and and align effect sizes to correspond to the same allele.
# It will remove variants with ambiguous alleles (G/C or A/T) or with alleles that do not match between data sets
# (e.g A/G in one data set and A/C in the other). It will also remove variants that are duplicated in either data set.
# It will not remove variants that are simply strand flipped between the two data sets (e. g. A/C in one data set, T/G in the other)

#Some of the outcome datasets didn't have a rsID/SNP column. Thus, we have to merge using variant ID. Create a variant ID column in X1 (diet data) that matches the X2 format (chr:bp:other:effect)
X1$variant_id <- paste0(X1$CHR, ":", X1$BP, ":", X1$ALLELE0, ":", X1$ALLELE1)
# Also annoying, in this outcome dataset (CVD) the alleles are lowercase
X2$effect_allele.outcome <- toupper(X2$effect_allele.outcome)
X2$other_allele.outcome <- toupper(X2$other_allele.outcome)
# Recreate variant id in X2 
X2$variant <- paste0(X2$chr.outcome, ":", X2$pos.outcome, ":", X2$other_allele.outcome, ":", X2$effect_allele.outcome)


X <- cause::gwas_merge(X1, X2, snp_name_cols = c("variant_id", "variant"), 
                       beta_hat_cols = c("BETA", "beta.outcome"), 
                       se_cols = c("SE", "se.outcome"), 
                       A1_cols = c("ALLELE1", "effect_allele.outcome"), # EFFECT ALLELE, per vignette example
                       A2_cols = c("ALLELE0", "other_allele.outcome"))
#pval_cols = c("P_BOLT_LMM", "LP"))
```

Run this for outcomes missing SNP/rsID (CVD)
```{r Deal with missing rsID}
X1snp <- X1 %>% select(c(SNP, variant_id))
X_snp <- merge(X, X1snp, by.x = "snp", by.y = "variant_id")
# Rename
X_snp <- X_snp %>% rename("variant_id" = "snp",
                          "snp" = "SNP")
```

```{r Step 1.5 - Save for LD pruning}
# Step 1.5: Save X for LD pruning on Alpine
saveRDS(X_snp, paste0("/pl/active/colelab/users/kjames/enviroMR/interim_data/MRCAUSE/", exposure, "_", outcome, "_nonLDpruned.rds"))

# Load when needed
X <- readRDS(paste0("/pl/active/colelab/users/kjames/enviroMR/interim_data/MRCAUSE/", exposure, "_", outcome, "_nonLDpruned.rds"))
```


```{r Step 2 - Calculate nuisance parameters}
# Step 2: Calculate nuisance parameters
# Takes several minutes
set.seed(100)
varlist <- with(X, sample(snp, size=1000000, replace=FALSE))
params <- cause::est_cause_params(X, varlist)

# Save for future
save(list=c("params"), file = paste0("/pl/active/colelab/users/kjames/enviroMR/interim_data/MRCAUSE/", exposure, "_", outcome, "_nonLDpruned_params.Rdata"))

# Load if needed
load(file = paste0("/pl/active/colelab/users/kjames/enviroMR/interim_data/MRCAUSE/", exposure, "_", outcome, "_nonLDpruned_params.Rdata"))
```

Step 3: LD Pruning
*THIS STEP REQUIRES YOU TO RUN SCRIPTS OUTSIDE OF THIS MARKDOWN*
See LDprune_MRCAUSE.R/.sh, and combine_LDprunedvariants.R
Download the output and load in next chunk.

```{r Step 3.5 - Prune variants}
# Load pruned variants
variants_pruned <- read.csv(paste0("/pl/active/colelab/users/kjames/enviroMR/MR_CAUSE/pruned_data/LD_1KG_pruned_",exposure,"_",outcome,".csv"))
variants_pruned <- variants_pruned[-1,3, drop = FALSE]

# Prune X
Xp <- X[X$snp %in% variants_pruned$V2,]
dim(Xp) #2358

# Save pruned variants
save(list=c("Xp"), file = paste0("/pl/active/colelab/users/kjames/enviroMR/interim_data/MRCAUSE/", exposure, "_", outcome, "_LDpruned_variants.Rdata"))
```

```{r Step 4 - Fit CAUSE model}
# Step 4: Fit CAUSE
res <- cause(X=X, variants = Xp$snp, param_ests = params) 

# Save - results are going into their specific exposure-outcome folder
save(list=c("res"), file = paste0("/pl/active/colelab/users/kjames/enviroMR/results/julie_collab/MRCAUSE/", exposure, "_", outcome,"/MRCAUSE_res_", exposure, "_", outcome, ".Rdata"))

# Load if needed
load(file = paste0("/pl/active/colelab/users/kjames/enviroMR/results/julie_collab/MRCAUSE/",exposure, "_", outcome,"/MRCAUSE_res_", exposure, "_", outcome, ".Rdata"))

```

Using res
```{r Step 5 - Review MR results}
# Step 5: Look at results
# model 1, model2: the models being compared
# delta_elpd: estimated difference in elpd. if negative, model 2 is a better fit.
# se_delta_elpd: estimated standard error of delta_elpd
# z: delta_elpd/se_delta_elpd: a z-score that can be compared to a normal distribution to test if the difference in model fit is significant

# Look at model statistics
# Sharing model
res$loos[[2]]

# Causal model
res$loos[[3]]

# Look at results
mr_elpd_res <- res$elpd # here sharing model is better (lower delta_elpd)

# Add pvalues to mr_elpd_res
mr_elpd_res$pvalue <- pnorm(mr_elpd_res$z, lower.tail=TRUE)
mr_elpd_res

res_list <- summary(res, ci_size=0.95)
res_table <- res_list$tab

# Multi-image plot from package
plot(res)
# Model plot from pacakge
plot(res, type = "data")

# Save
# USER INPUT: Check paths and variable names
# Set path
res_dir_mrcause = paste0("/pl/active/colelab/users/kjames/enviroMR/results/julie_collab/MRCAUSE/", exposure, "_", outcome)
# Save tables
write.csv(mr_elpd_res, file = paste0(res_dir_mrcause, "/mrcause_res_elpd_", exposure, "_", outcome, ".csv"))
write.csv(res_table, file = paste0(res_dir_mrcause, "/mrcause_res_table_", exposure, "_", outcome, ".csv"))
# Save plots
png(paste0(res_dir_mrcause,"/mrcause_plotres_",exposure, "_", outcome,"_results.png"), width = 900, height = 600)
plot(res)
dev.off()

png(paste0(res_dir_mrcause,"/mrcause_plotdata_",exposure, "_", outcome,"_results.png"), width = 900, height = 600)
plot(res, type = "data")
dev.off()
```
Using res_5e8
```{r Step 5 - Review MR results}
# Step 5: Look at results
# model 1, model2: the models being compared
# delta_elpd: estimated difference in elpd. if negative, model 2 is a better fit.
# se_delta_elpd: estimated standard error of delta_elpd
# z: delta_elpd/se_delta_elpd: a z-score that can be compared to a normal distribution to test if the difference in model fit is significant

# Look at model statistics
# Sharing model
res_5e8$loos[[2]]

# Causal model
res_5e8$loos[[3]]

# Look at res_5e8ults
mr_elpd_res_5e8 <- res_5e8$elpd # here sharing model is better (lower delta_elpd)

# Add pvalues to mr_elpd_res_5e8
mr_elpd_res_5e8$pvalue <- pnorm(mr_elpd_res_5e8$z, lower.tail=TRUE)
mr_elpd_res_5e8

res_5e8_list <- summary(res_5e8, ci_size=0.95)
res_5e8_table <- res_5e8_list$tab

# Multi-image plot from package
plot(res_5e8)
# Model plot from pacakge
plot(res_5e8, type = "data")

# Save
# USER INPUT: Check paths and variable names
# Set path
#res_5e8_dir_mrcause = paste0("/pl/active/colelab/users/kjames/enviroMR/results/julie_collab/MRCAUSE/", exposure, "_", outcome)
# Save tables
write.csv(mr_elpd_res_5e8, file = paste0(res_dir_mrcause, "/mrcause_res5e8_elpd_", exposure, "_", outcome, ".csv"))
write.csv(res_5e8_table, file = paste0(res_dir_mrcause, "/mrcause_res5e8_table_", exposure, "_", outcome, ".csv"))
# Save plots
#png(paste0(res_dir_mrcause,"/mrcause_plotres_",exposure, "_", outcome,"_results.png"), width = 900, height = 600)
#plot(res)
#dev.off()

#png(paste0(res_dir_mrcause,"/mrcause_plotdata_",exposure, "_", outcome,"_results.png"), width = 900, height = 600)
#plot(res, type = "data")
#dev.off()
```

Testing plotting techniques below. Safe to ignore.
```{r}
head(Xp)

custom_MRscatter_std <- ggplot(Xp, aes(x = beta_hat_1, y = beta_hat_2)) +
  geom_point() +
  #outcome
  #ggplot2::geom_errorbar(ggplot2::aes(ymin=beta_hat_2-seb2, ymax=beta_hat_2+seb2), colour="grey", width=0) +
  #exposure
  #ggplot2::geom_errorbarh(ggplot2::aes(xmin=beta_hat_1-seb1, xmax=beta_hat_1+seb1), colour="grey", height=0) +
  #line
  #ggplot2::geom_abline(data = mr_res_df_subsetMethods, # Specify correct data here
   #                    ggplot2::aes(intercept=0, slope=Estimate, colour=Method), size = 1.75, show.legend=TRUE) + #0 if not MR-Egger
  ggplot2::theme_bw() +
  ggplot2::theme(
    #legend.position="none", 
    axis.title.y=ggplot2::element_text(size=14),
    axis.text.y=ggplot2::element_text(size=14), 
    axis.ticks.y=ggplot2::element_line(size=0),
    axis.title.x=ggplot2::element_text(size=14),
    axis.text.x=ggplot2::element_text(size=14),
    axis.ticks.x=ggplot2::element_line(size=0),
    panel.border = element_rect(color = "black", 
                                fill = NA, 
                                size = 3)) +
  labs(x = paste0("Genetic Effect on ", EXP_name),
       y = paste0( "Genetic Effect on ", OUT_name)) +
  ggtitle(paste0("Non-filtered Genetic Instrument (n=", nrow(Xp), ")")) +
  scale_color_manual(values = c("#3E4A89FF", "#6DCD59FF"))

custom_MRscatter_std
```

